src/transpiling/jsx_precompile.rs

```diff
@@ -0,0 +1,1825 @@
+    dynamic_exprs: &mut Vec<Expr>,
+  ) {
+    let name: &str = match &el.opening.name {
+      // Case: <div />
+      JSXElementName::Ident(ident) => &ident.sym,
+      _ => {
+        unreachable!("serialize_jsx_element_to_string_vec(JSXNamespacedName)")
+      }
+    };
+
+    // Case: <div {...props} />
+    // Case: <div class="foo" {...{ class: "bar"}} />
+    // Case: <div {...{ class: "foo"}} class="bar"}>foo</div>
+    // Case: <Foo />
+    if !is_serializable(&el.opening) {
+      let expr = Expr::Call(self.serialize_jsx_to_call_expr(el));
+      strings.push("".to_string());
+      dynamic_exprs.push(expr);
+
+      return;
+    } else if strings.is_empty() {
+      strings.push("".to_string());
+    }
+
+    strings.last_mut().unwrap().push('<');
+
+    let escaped_name = escape_html(name);
+    strings.last_mut().unwrap().push_str(escaped_name.as_str());
+
+    for attr in &el.opening.attrs {
+      // Case: <button class="btn">
+      match attr {
+        JSXAttrOrSpread::JSXAttr(jsx_attr) => {
+          let attr_name = get_attr_name(jsx_attr);
+
+          // Case: <input required />
+          let Some(attr_value) = &jsx_attr.value else {
+            strings.last_mut().unwrap().push(' ');
+            let escaped_attr_name = escape_html(&attr_name);
+            strings
+              .last_mut()
+              .unwrap()
+              .push_str(escaped_attr_name.as_str());
+            continue;
+          };
+
+          // Case: <div class="btn">
+          // Case: <div class={"foo"}>
+          // Case: <div class={2}>
+          // Case: <div class={true}>
+          // Case: <div class={null}>
+          match attr_value {
+            JSXAttrValue::Lit(lit) => match lit {
+              Lit::Str(string_lit) => {
+                // Edge Case: Both "key" and "ref" attributes are
+                // special attributes in most frameworks. Some
+                // frameworks may want to serialize it, other's don't.
+                // To support both use cases we'll always pass them to
+                // `jsxattr()` so that frameowrks can decide for
+                // themselves what to do with it.
+                // Case: <div key="123" />
+                // Case: <div ref="123" />
+                if attr_name == "key" || attr_name == "ref" {
+                  strings.last_mut().unwrap().push(' ');
+                  strings.push("".to_string());
+                  let expr = self.convert_to_jsx_attr_call(
+                    attr_name,
+                    string_lit_expr(string_lit.value.to_string()),
+                  );
+                  dynamic_exprs.push(Expr::Call(expr));
+                  continue;
+                }
+
+                let serialized_attr = format!(
+                  " {}=\"{}\"",
+                  escape_html(&attr_name).as_str(),
+                  escape_html(string_lit.value.as_ref()).as_str()
+                );
+
+                strings.last_mut().unwrap().push_str("");
```

Comment:
Seems this can be removed or it's a bug? If it's a bug, perhaps we should add a test?